### PR TITLE
Cookie-based auth

### PR DESCRIPTION
**Pull request by dokterbob**
Implementation of #1520

- [x] Overall auth cleanup/refactor
- [x] Basic implementation
- [x] E2E-tests passing
- [x] Initial cleanup
- [ ] Manual testing with OAuth
- [ ] Basic copilot integration
- [ ] Manual or automated testing of copilot
- [ ] Unit tests for auth methods
- [ ] Create ticket for serving files from different root path (out of scope for this one, see below)
- [ ] Create ticket for using [OAuth Authcode](https://www.oauth.com/oauth2-servers/server-side-apps/authorization-code/) flow copilot

### CODE REVIEW

**Comment by dokterbob** [comment]
Just 2 more tests missing:
```
────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  header_auth/spec.cy.ts                                                        (18 of 37)


  Header auth
    1) should fail to auth without custom header
    ✓ should be able to auth with custom header (6430ms)


  1 passing (16s)
  1 failing

  1) Header auth
       should fail to auth without custom header:
     CypressError: The application redirected to `http://127.0.0.1:8000/login` more than 20 times. Please check if it's an intended behavior.

If so, increase `redirectionLimit` value in configuration.
      at onWindowLoad (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:137531:80)
      at $Cy.onInternalWindowLoad (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:137563:20)
      at $Cy.listener (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:86503:17)
      at ../driver/node_modules/eventemitter2/lib/eventemitter2.js.EventEmitter.emit (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:86583:19)
      at parent.<computed> [as emit] (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:155168:32)
      at $Cypress.action (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:147734:14)
      at HTMLIFrameElement.<anonymous> (http://127.0.0.1:8000/__cypress/runner/cypress_runner.js:151065:24)
  From Your Spec Code:
      at Context.eval (webpack:///./cypress/e2e/header_auth/spec.cy.ts:9:0)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        2                                                                                │
  │ Passing:      1                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     15 seconds                                                                       │
  │ Spec Ran:     header_auth/spec.cy.ts                                                           │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /home/runner/work/chainlit/chainlit/cypress/screenshots/header_auth/spec.cy.ts/H     (1280x720)
     eader auth -- should fail to auth without custom header (failed).png                           

────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  data_layer/spec.cy.ts                                                         (12 of 37)


  Data Layer
    Data Features with persistence
      1) should login, submit feedback, wait for user input to create steps, browse thread history, delete a thread and then resume a thread
      2) should continue the thread after backend restarts and work with new thread as usual


  0 passing (41s)
  2 failing

  1) Data Layer
       Data Features with persistence
         should login, submit feedback, wait for user input to create steps, browse thread history, delete a thread and then resume a thread:
     AssertionError: Timed out retrying after 15000ms: expected '/' to match /^\/thread\//
      at Context.eval (webpack:///./cypress/e2e/data_layer/spec.cy.ts:16:0)

  2) Data Layer
       Data Features with persistence
         should continue the thread after backend restarts and work with new thread as usual:
     AssertionError: Timed out retrying after 15000ms: expected '/' to match /^\/thread\//
      at Context.eval (webpack:///./cypress/e2e/data_layer/spec.cy.ts:16:0)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        2                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      2                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      0                                                                                │
  │ Screenshots:  2                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     40 seconds                                                                       │
  │ Spec Ran:     data_layer/spec.cy.ts                                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /home/runner/work/chainlit/chainlit/cypress/screenshots/data_layer/spec.cy.ts/Da     (1280x720)
     ta Layer -- Data Features with persistence -- should login, submit feedback, wai               
     t for user input to create steps, browse thread history, delete a thread and the               
     n resume a thread (failed).png                                                                 
  -  /home/runner/work/chainlit/chainlit/cypress/screenshots/data_layer/spec.cy.ts/Da     (1280x720)
     ta Layer -- Data Features with persistence -- should continue the thread after b               
     ackend restarts and work with new thread as usual (failed).png                                 


────────────────────────────────────────────────────────────────────────────────────────────────────```

**Comment by dokterbob** [comment]
Anyone a Windows box lying around? 🤯 

image

**Comment by dokterbob** [comment]
@hmrc87 @celeriev Would love initial feedback. Might have been stuff I missed, kind of a large chunk of work. ;)

**Comment by dokterbob** [comment]
# Copilot auth
## Option A
Simplified OAuth, assuming that the owner of client_secret can be trusted. JWT token contains client_id and is signed by client_secret. Server needs to be configured to allow client_id, which is used to authenticate client_secret. JWT exchange token is very short lived (minutes).

image

Here's a roll-up of the authentication system:

### Secure Embedded App Authentication

This system enables secure authentication for embedded applications using Ed25519 public-key cryptography and HTTPOnly cookies, balancing security with a simple developer experience.

#### Overview

1. Each integration partner gets:
   - A private key (client_secret)
   - A corresponding public key (serves as client_id)

2. Authentication Flow:
   - Host server generates a JWT containing:
     - User identity/permissions
     - Their public key
     - Signed with their private key
   - Embedded app initializes with this JWT
   - API validates JWT using public key from the token
   - Upon validation, sets secure HTTPOnly cookie
   - Further requests use cookie auth

#### Security Properties

- Asymmetric crypto (Ed25519) ensures hosts can only sign for themselves
- Self-contained tokens (pubkey included) with minimal server config
- Server whitelists allowed client_ids for access control
- HTTPOnly cookies prevent XSS after initial auth
- 256-bit Ed25519 keys provide strong security with small size

#### Key Benefits

- Zero server config for new integrations (beyond client_id whitelist)
- Clean developer experience
- Single auth exchange at session start
- Clear audit trail (client_id identifies signer)
- Strong security model with minimal complexity

#### Limitations

- Initial JWT briefly exposed to JavaScript
- Requires hosts to securely store private key
- Need to maintain client_id whitelist

## Option B
Server calling out to copilot uses OAuth to authenticate with chainlit, chainlit generates a short-lived refresh token, server shares it with the client, client exchanges for a cookie.

This is the most standardised but also adds a considerable amount of complexity.

The key points:

* Host server uses client credentials + custom claims for user context
* Short-lived initial token only used once to establish cookie
* After exchange, everything works with HTTPOnly cookies
* User never sees auth process
* Maintains security while being invisible to end user

image

**Comment by dokterbob** [comment]
I hugely appreciate community feedback and/or a security review of the aforementioned options for cookie auth. @hmrc87 @celeriev @willydouhard @robkrause

File: backend/chainlit/auth/cookie.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        self.scheme_name = scheme_name or self.__class__.__name__
+        self.auto_error = auto_error
+
+    async def __call__(self, request: Request) -> Optional[str]:
+        # First try to get the token from the cookie
+        token = request.cookies.get(_auth_cookie_name)
+
+        # If no cookie, try the Authorization header as fallback
+        if not token:
+            # TODO: Only bother to check if cookie auth is explicitly disabled.
```
Comment (willydouhard) [inline backend/chainlit/auth/cookie.py:53]: are we leaving this todo?

File: backend/chainlit/server.py [hunk 2]
```diff
@@ -363,43 +370,109 @@ async def auth(request: Request):
...
+    if redirect_to_callback:
+        root_path = os.environ.get("CHAINLIT_ROOT_PATH", "")
+        redirect_url = (
+            f"{root_path}/login/callback?{urllib.parse.urlencode(response_dict)}"
         )
 
-    user = await config.code.password_auth_callback(
-        form_data.username, form_data.password
+        return RedirectResponse(
+            # FIXME: redirect to the right frontend base url to improve the dev environment
```
Comment (willydouhard) [inline backend/chainlit/server.py:398]: should we fix this before RC?

**Pull request merged**